### PR TITLE
return None when specified version not found in git.get_version

### PR DESCRIPTION
**Pull request by wkentaro**
to Fix: https://github.com/vcstools/wstool/issues/41

### CODE REVIEW

File: src/vcstools/common.py [hunk 1]
```diff
@@ -340,7 +340,8 @@ def run_shell_command(cmd, cwd=None, shell=False, us_env=True,
             if cwd is not None:
                 message += "\n run at: '%s'" % (cwd)
             message += "\n errcode: %s:\n%s" % (proc.returncode, stderr)
-            logger.warn(message)
+            if verbose:
```
Comment (tkruse) [inline src/vcstools/common.py]: This is a very big change. Some wstool users want to see certain messages. So I think we cannot do it like that. Maybe we will need the no_warn argument back...

Comment (wkentaro) [inline src/vcstools/common.py]: Okay, for historical reason, I revert no_warn.

File: src/vcstools/git.py [hunk 2]
```diff
@@ -368,12 +368,16 @@ def get_version(self, spec=None):
...
-                                                 shell=True,
-                                                 cwd=self._path)
+            for i in range(2):
+                retcode, output, _ = run_shell_command(command,
+                                                       shell=True,
+                                                       no_warn=True,
+                                                       cwd=self._path)
                 if (output != '' or spec is None):
                     break
+                if i == 1 and retcode != 0 and spec is not None:
```
Comment (tkruse) [inline src/vcstools/git.py]: do we need the returncode at all? Maybe there should just be if i == 0 before the fetch

Comment (wkentaro) [inline src/vcstools/git.py]: returncode is for returning `None` not `''`
if i == 0 before the fetch is just a refactoring for useless fetching at second time.

**Comment by tkruse** [comment]
You seem to still be working on this, please add a comment when you want another review.
A testcase would be nice, and probably not too difficult (I just mean testing return of None.)

**Comment by wkentaro** [comment]
sorry about that. I added a test case for not existing version.
And I'm done, please review this.

File: src/vcstools/git.py [hunk 3]
```diff
@@ -366,22 +366,28 @@ def get_version(self, spec=None):
...
+            _, output, _ = run_shell_command(command, shell=True,
+                                             no_warn=True, cwd=self._path)
+            if output.strip() == '':
+                # even if after fetching, not found specified version
+                return None
+            version = output.strip()
             # On Windows the version can have single quotes around it
-            output = output.strip("'")
-            return output
+            version = version.rstrip("'").lstrip("'")
```
Comment (tkruse) [inline src/vcstools/git.py]: what was wrong with strip("'") ?

File: src/vcstools/git.py [hunk 4]
```diff
@@ -366,22 +366,28 @@ def get_version(self, spec=None):
...
-                # we try again after fetching if given spec had not been found
-                try:
-                    self._do_fetch()
-                except GitError:
-                    return None
+            _, output, _ = run_shell_command(command, shell=True,
+                                             no_warn=True, cwd=self._path)
+            if output.strip() != '':
+                version = output.strip()
+                return version  # found SHA-ID
```
Comment (tkruse) [inline src/vcstools/git.py]: not stripping quotes here

**Comment by wkentaro** [comment]
thanks, fixed.

**Comment by tkruse** [comment]
OK, code looks good to me, maybe commits could be squashed down to 3 or 4, refactorings separate, the rest in one commit.

**Comment by tkruse** [comment]
Also the commit message should probably say that what this actually fixes is dirty output for wstool info in case a specified branch does not exist. @wjwwood consider merging when travis is happy.

**Comment by wkentaro** [comment]
Rebased, and please review commit message:
https://github.com/wkentaro/vcstools/commit/6a18cd3d4c503e1b5a502e82cfa5002e46b527b0

**Comment by wkentaro** [comment]
squashed again, so please see commit message here:
https://github.com/vcstools/vcstools/commit/3b64279c9f393a9997c086eb209bd4b05c185391

**Comment by tkruse** [comment]
very nice, thank you. Also, in the buildsystem SIG you could still answer my last post about tracked branches in your team.

**Comment by wjwwood** [comment]
lgtm, waiting on CI to catch up.

**Pull request merged**